### DISCUSSION TITLE
Unsupported tool chain (CCRX)

### DISCUSSION THREAD

**Discussion by Wini-Buh** [hathach/tinyusb#785] [Q&A]
Hi

I found your project after searching for a solution of a easy to use USB stack. After looking at the supported MCUs, I was surprised to see the RX63N from Renesas in the list.
I use the RX72N and the CCRX tool chain from Renesas. Unfortunately this tool chain it not support, so I decided trying to add this support to the project. I modified the file "tusb_compiler.h" accordingly to my needs. One problem I could not solve easily was the missing support of the weak functionality, which is not available in the tool chain from Renesas. To solve this problem I decided to add a "weak wrapper" file, which (hopefully) should solve this.

My question is now... Is there any interest to add this changes to the project, if they are working ?

BTW: It was also required to make changes for the "packed attribute", because the CCRX tool chain only support it trough a #pragma directive.  So I decided to use an approach, which is used in the lwIP TCP/IP stack to make it portable.
 
Thank you very much for an answer.

**Comment by hathach** [comment]
Hi,

> I found your project after searching for a solution of a easy to use USB stack. After looking at the supported MCUs, I was surprised to see the RX63N from Renesas in the list.

Yeah, the rx63n support is only recently added https://github.com/hathach/tinyusb/pull/757

> I use the RX72N and the CCRX tool chain from Renesas. Unfortunately this tool chain it not support, so I decided trying to add this support to the project. I modified the file "tusb_compiler.h" accordingly to my needs. One problem I could not solve easily was the missing support of the weak functionality, which is not available in the tool chain from Renesas. To solve this problem I decided to add a "weak wrapper" file, which (hopefully) should solve this.

> My question is now... Is there any interest to add this changes to the project, if they are working ?

I am open to all suggestions if it fits. Since you already done the work. Would you mind submit an PR for it, it doesn't have to work, just to see how much of the changes you are proposing. Also maybe others could join the discussion as well.
 
> BTW: It was also required to make changes for the "packed attribute", because the CCRX tool chain only support it trough a #pragma directive. So I decided to use an approach, which is used in the lwIP TCP/IP stack to make it portable.

can you declare it with `_Pragma()` in place of the ATTR_PACKED ? Would you mind giving the link to the way lwip is doing. Again I am open to all suggestion if it makes sense.  

> Thank you very much for an answer.

No problem, I am the one to say thank

**Reply by Wini-Buh** [reply]
> > I use the RX72N and the CCRX tool chain from Renesas. Unfortunately this tool chain it not support, so I decided trying to add this support to the project. I modified the file "tusb_compiler.h" accordingly to my needs. One problem I could not solve easily was the missing support of the weak functionality, which is not available in the tool chain from Renesas. To solve this problem I decided to add a "weak wrapper" file, which (hopefully) should solve this.
> 
> > My question is now... Is there any interest to add this changes to the project, if they are working ?
> 
> I am open to all suggestions if it fits. Since you already done the work. Would you mind submit an PR for it, it doesn't have to work, just to see how much of the changes you are proposing. Also maybe others could join the discussion as well.

Ok, I will do that. But first, I prefer to have a running implementation of tinyUSB (which is not the case at the moment).
> 
> > BTW: It was also required to make changes for the "packed attribute", because the CCRX tool chain only support it trough a #pragma directive. So I decided to use an approach, which is used in the lwIP TCP/IP stack to make it portable.
> 
> can you declare it with `_Pragma()` in place of the ATTR_PACKED ? Would you mind giving the link to the way lwip is doing. Again I am open to all suggestion if it makes sense.

 Here are two links to files to the lwIP repo :

https://git.savannah.nongnu.org/cgit/lwip.git/tree/src/include/lwip/arch.h
This file defines the required "packed macros" to implement the "packed structs"

https://git.savannah.nongnu.org/cgit/lwip.git/tree/src/include/lwip/prot/icmp.h
This file shows some example how the "packed macros" are used

For my implementation I only used the macros PACK_STRUCT_BEGIN and PACK_STRUCT_END, like these :
`#define TU_PACK_STRUCT_BEGIN _Pragma("pack")`
`#define TU_PACK_STRUCT_END _Pragma("packoption")`

For example this looks like this (struct found in class/cdc/cdc.h in the tinyUSB project):

`TU_PACK_STRUCT_BEGIN`
`typedef struct TU_ATTR_PACKED`
`{`
`  uint8_t bLength                    ; ///< Size of this descriptor in bytes.`
`  uint8_t bDescriptorType       ; ///< Descriptor Type, must be Class-Specific`
`  uint8_t bDescriptorSubType ; ///< Descriptor SubType one of above CDC_FUNC_DESC_`
`  uint16_t bcdCDC                  ; ///< CDC release number in Binary-Coded Decimal`
`}cdc_desc_func_header_t;`
`TU_PACK_STRUCT_END`

The definition of the macro `TU_ATTR_PACKED` is left empty.

I hope these gives you an idea how I use it.

**Reply by hathach** [reply]
> Ok, I will do that. But first, I prefer to have a running implementation of tinyUSB (which is not the case at the moment).

It shouldn't be an issue, I only want to have a peek at what the changes to go around the weak function. There is similar issue with segger linker that is fixed by explicitly enabling an linker option `--undefined-weak-is-zero`  https://github.com/hathach/tinyusb/issues/513 . Maybe you could check/ask this on renesas forum, weak symbol is very common in library implmentation e.g libc. I am hoping we could support ccrx without too much of code changes.

Thank you for the detail explanation and example, I see how the packed work with. It isn't that bad, since most of the packed struct is normally defined together. can we have multiple packed struct in a pair of begin end as follow. 

```C
TU_PACK_STRUCT_BEGIN
typedef struct TU_ATTR_PACKED
{
  uint8_t bLength            ; ///< Size of this descriptor in bytes.
} midi_desc_header_t;

/// MIDI In Jack Descriptor
typedef struct TU_ATTR_PACKED
{
  uint8_t bLength            ; ///< Size of this descriptor in bytes.
} midi_desc_in_jack_t;

/// MIDI Out Jack Descriptor with single pin
typedef struct TU_ATTR_PACKED
{
  uint8_t bLength            ; ///< Size of this descriptor in bytes.
} midi_desc_out_jack_t ;

TU_PACK_STRUCT_END
``` 

There is another concern is align, would CCRX support struct member alignment. Since I remember having similar issue with IAR before they support gcc attribute. 

```C
struct TU_ATTR_PACKED {
  uint8_t a;
  TU_ATTR_ALGINED(4)  uint8_t b4;
} my_struct;
```

**Reply by Wini-Buh** [reply]
> Thank you for the detail explanation and example, I see how the packed work with. It isn't that bad, since most of the packed struct is normally defined together. can we have multiple packed struct in a pair of begin end as follow.

That's correct. The default pack setting is overridden by the  `_Pragma()` until another `_Pragma()` with a pack directive is found (or the end of the C file is reached, if I remember correctly).

> There is another concern is align, would CCRX support struct member alignment. Since I remember having similar issue with IAR before they support gcc attribute.

No, that's not supported. Alignment must be done manually. How did you overcome the problem with the IAR tool chain before they supported the alignment attribute ?
BTW: I also recognized the problem while trying porting the tiny USB to the CCRX tool chain.

I also found some other issue with the bit fields. That is, the orientation of a bit field is implementation dependent, that means there is no standard where the MSB/LSB in a bit field reside (left or right)... I have solved this with an additionally `_Pragma()`.

**Reply by hathach** [reply]
> That's correct. The default pack setting is overridden by the `_Pragma()` until another `_Pragma()` with a pack directive is found (or the end of the C file is reached, if I remember correctly).

thanks for confirmation, that will make thing easier to port
 
> No, that's not supported. Alignment must be done manually. How did you overcome the problem with the IAR tool chain before they supported the alignment attribute ?
> BTW: I also recognized the problem while trying porting the tiny USB to the CCRX tool chain.

I have never fixed issue with IAR, I have initial port for IAR, but gave up due to amount of works, and almost no body here use it. It is added recently when IAR support gcc attribute. Btw, in-struct alignment is not a deal breaker. It is used mostly (for now) for NXP LPC mcus where they require USB buffer must be aligned 64 bytes. If Renesas MCU does not have such requirement, I don't see any issue with  that. I still have no idea if IAR support in-struct alignment since NXP LPC usually is used with gcc. 
 
> I also found some other issue with the bit fields. That is, the orientation of a bit field is implementation dependent, that means there is no standard where the MSB/LSB in a bit field reside (left or right)... I have solved this with an additionally `_Pragma()`.

That is right, but the gcc order is also most the universal (default) order. Please try to see if by default, it would match with gcc implementation.

**Reply by hathach** [reply]
Before continue with the compiler port, maybe you should try to use the gccrx with existing example first. You may need to do a bit of board porting. Best to verify your board with the working bit first before expanding it.

**Reply by Wini-Buh** [reply]
Success... the port seems to work (of course it's not sure it works on every condition, but it's a first step). I have the CDC device working and a small echo task is answering every incoming data.

I had a few pitfalls to overcome... For the actual project with the Rx72N, I use a lot of old source code written for the Renesas SH2A CPU (it is a RISC 32 bit uController). This CPU runs in big endian mode (first pitfall, because we use this endian mode also on the Rx72N) and uses bit field ordering from the left, means bit field members are assigned from the upper-bit side (second pitfall, same reason as on the endian mode). The tool chain I use is actually not supported, means the attribute directive is not supported, etc. (third pitfall).

**Reply by hathach** [reply]
wow, great work, I am surprised that you could pull this off with the big-endian MCU. I have never tested this with BE MCU before. Did the MSC example work, MSC has some endian related data, you may need to check/port the endian swap macro tu_htonl() to get it work.

**Reply by Wini-Buh** [reply]
No, I did not test it with MSC. I could test it (and make the required modification for it), but I need to do that in my spare time.
BTW: The USB HW-IP of the RX CPU has a feature to access the FIFOs in little or big endian mode, so this solved me a lot of byte swapping while accessing the FIFOs.

Regarding the tu_htonl() (and the other macros)... I have already adapted it to my port.

Some other word about the tinyUSB. I'm really positively surprised about the tinyUSB, the source code itself and very good debugging support... great work from you. I first tried to use the USB stack from Renesas, but from my point of view this code is a real catastrophe and the documentation also.

**Reply by hathach** [reply]
thank you for your kind words, renesas is still in early stage. No pressure at all with msc testing, I work on TinyUSB stack in my spare time as well. Let just keep up your great work and post your progress and/or problems here as well.

**Comment by abakosh** [comment]
I have the same issue with IAR packed structure.  Please refer to the following for IAR support 

![Screenshot_20230512_221401_Chrome.jpg](https://github.com/hathach/tinyusb/assets/36272850/f1310428-75c4-443a-a065-3c418c0d5e73)

I was thinking about implementing __packed for RA family, which is based on ARM. Unfortunately, it is not supported for all the architecture of IAR

I will do my suggested changes for RA and send you a PR
